ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaSource.java

```diff
@@ -0,0 +1,215 @@
+      EnumSet.of(QuotaResourceType.ACCOUNT, QuotaResourceType.CONTAINER);
+  private static final Logger LOGGER = LoggerFactory.getLogger(AmbryCUQuotaSource.class);
+  private static final long DEFAULT_RCU_FOR_NEW_RESOURCE = 0;
+  private static final long DEFAULT_WCU_FOR_NEW_RESOURCE = 0;
+  protected final CapacityUnit feQuota; // Ambry frontend's CU capacity.
+  private final ConcurrentMap<String, CapacityUnit> cuQuota; // in memory quota for all resources.
+  private final ConcurrentMap<String, CapacityUnit> cuUsage; // in memory quota usage for all resources.
+  private final ScheduledExecutorService usageRefresher;
+  private final long aggregationWindowsInSecs;
+  private final AtomicBoolean isReady;
+  protected final AtomicReference<CapacityUnit> feUsage; // Ambry frontend's CU usage.
+
+  /**
+   * Constructor for {@link AmbryCUQuotaSource}.
+   * @param quotaConfig {@link QuotaConfig} object.
+   * @param accountService {@link AccountService} object.
+   * @throws IOException in case of any exception.
+   */
+  public AmbryCUQuotaSource(QuotaConfig quotaConfig, AccountService accountService) throws IOException {
+    feQuota = JsonCUQuotaDataProviderUtil.getFeCUCapacityFromJson(quotaConfig.frontendCUCapacityInJson);
+    cuQuota = new ConcurrentHashMap<>(
+        JsonCUQuotaDataProviderUtil.getCUQuotasFromJson(quotaConfig.resourceCUQuotaInJson, accountService));
+    cuUsage = new ConcurrentHashMap<>();
+    usageRefresher = Utils.newScheduler(1, REFRESHER_THREAD_NAME_PREFIX, true);
+    aggregationWindowsInSecs = quotaConfig.cuQuotaAggregationWindowInSecs;
+    isReady = new AtomicBoolean(false);
+    feUsage = new AtomicReference<>(null);
+  }
+
+  @Override
+  public synchronized void init() {
+    if (isReady.get()) {
+      LOGGER.warn("InMemoryCuQuotaSource is already initialized.");
+      return;
+    }
+    feUsage.set(new CapacityUnit());
+    cuQuota.keySet().forEach(key -> cuUsage.put(key, new CapacityUnit()));
+    usageRefresher.scheduleAtFixedRate(this::resetQuotaUsage, aggregationWindowsInSecs, aggregationWindowsInSecs,
+        TimeUnit.SECONDS);
+    isReady.set(true);
+  }
+
+  @Override
+  public boolean isReady() {
+    return isReady.get();
+  }
+
+  @Override
+  public Quota getQuota(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+
+    final String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    return new Quota<>(quotaName, cuQuota.get(resourceId).getQuotaValue(quotaName), quotaResource);
+  }
+
+  @Override
+  public float getUsage(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    final String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    double usage = 0;
+    if (cuUsage.containsKey(resourceId)) {
+      usage = cuUsage.get(resourceId).getQuotaValue(quotaName);
```

Comment:
This is a nit but, you can avoid looking up a value in a hashmap twice by doing:
```
CapacityUnit cu = cuUsage.get(resourceId);
double usage = cu != null? cu.getQuotaValue(quotaName) : 0;
```
(also could use optionals but we don't widely use them in our codebase)

Reply:
fixed.